core/client/src/notifications.rs

```diff
@@ -147,32 +208,84 @@ impl<Block: BlockT> StorageNotifications<Block> {
 						}
 					}
 				},
+			};
+			if let Some(child_filters) = child_filters {
+				for (c_key, filters) in child_filters.into_iter() {
+					if let Some((listeners, wildcards)) = self.child_listeners.get_mut(&c_key) {
+						if let Some(filters) = filters {
+							for key in filters.into_iter() {
+								let remove_key = match self.listeners.get_mut(&key) {
+									Some(ref mut set) => {
+										set.remove(&subscriber);
+										set.is_empty()
+									},
+									None => false,
+								};
+
+								if remove_key {
+									listeners.remove(&key);
+								}
+							}
+						} else {
+							wildcards.remove(&subscriber);
+						}
+						if listeners.is_empty() && wildcards.is_empty() {
+							self.child_listeners.remove(&c_key);
+						}
+					}
+				}
 			}
 		}
 	}
 
 	/// Start listening for particular storage keys.
-	pub fn listen(&mut self, filter_keys: Option<&[StorageKey]>) -> StorageEventStream<Block::Hash> {
+	pub fn listen(
+		&mut self,
+		filter_keys: Option<&[StorageKey]>,
+		filter_child_keys: Option<&[(StorageKey, Option<Vec<StorageKey>>)]>,
+	) -> StorageEventStream<Block::Hash> {
 		self.next_id += 1;
+		let next_id = self.next_id;
 
 		// add subscriber for every key
 		let keys = match filter_keys {
 			None => {
-				self.wildcard_listeners.insert(self.next_id);
+				self.wildcard_listeners.insert(next_id);
 				None
 			},
 			Some(keys) => Some(keys.iter().map(|key| {
 				self.listeners
 					.entry(key.clone())
 					.or_insert_with(Default::default)
-					.insert(self.next_id);
+					.insert(next_id);
 				key.clone()
 			}).collect())
 		};
+		let child_keys = filter_child_keys.map(|filter_child_keys| {
+			filter_child_keys.iter().map(|(c_key, o_keys)| {
+				let (c_listeners, c_wildcards) = self.child_listeners
+					.entry(c_key.clone())
+					.or_insert_with(Default::default);
+
+				(c_key.clone(), if let Some(keys) = o_keys {
```

Comment:
The code feels complicated here, I'd try to extract it similar as in `notify`, so that:
```
add_listeners(filter_keys, &mut self.wildcard_listeners, &mut self.listeners);
```
can be re-used for top-level keys and child keys